### PR TITLE
Node-level Checkpointing manager: Migrate dockershim and device plugin manager checkpointing  

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:
This PR abstracts checkpoint manager at kubelet level. Currently,  `dockershim`, `deviceplugin` have their own native checkpointing primitives. And most recently `cpumanager` also added package native checkpointing primitives. This adds to the redundancy at implementation level. Also degrades code readability and consistency.

To help this:

1.  Checkpointing interface is being abstracted at kubelet level as `checkpointmanager` package.
2.  `dockershim` and `deviceplugin` packages are modified to use `checkpointmanager` instead native checkpointing.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

```release-note
None
```
cc @jeremyeder @vishh @derekwaynecarr @sjenning @yujuhong @dchen1107 @RenaudWasTaken @ConnorDoyle @RenaudWasTaken @jiayingz @mindprince @timstclair 
/sig node

### CODE REVIEW

**Comment by ConnorDoyle** [comment]
cc @flyingcougar

**Comment by yujuhong** [comment]
Factoring the code out to be a reusable package definitely helps with the reliability.
This is quite a big change and we are already in code slush for 1.9. Is there any reason that this should be targeted for 1.9?

**Comment by tengqm** [comment]
#54088

**Comment by derekwaynecarr** [comment]
i am fine with this refactor landing post 1.9 unless there is an obvious bug with each existing implementation fixed by this.

**Comment by vikaschoudhary16** [comment]
Since this PR is not fixing any known issues, makes sense to me as well to target it post 1.9.
Meanwhile, reviews are welcome.

**Comment by yujuhong** [comment]
Adding @mtaufen, who has also written checkpointing code in kubelet
/assign @mtaufen 


> Since this PR is not fixing any known issues, makes sense to me as well to target it post 1.9.
> Meanwhile, reviews are welcome.

Sure. I think many people will not start reviewing until next week :-)

**Comment by sjenning** [comment]
@vikaschoudhary16 this needs an bazel update and gofmt

**Comment by sjenning** [comment]
deep in the logs, finally found
```kubelet[1629]: error: failed to run Kubelet: failed to create kubelet: failed to initialize checkpoint manager: mkdir : no such file or directory
```

**Comment by vikaschoudhary16** [comment]
Folks PTAL!

File: pkg/kubelet/checkpoint/checkpoint.go [hunk 1]
```diff
@@ -39,54 +36,57 @@ const (
...
-type Manager interface {
-	// LoadPods will load checkpointed Pods from disk
-	LoadPods() ([]*v1.Pod, error)
-
-	// WritePod will serialize a Pod to disk
-	WritePod(pod *v1.Pod) error
+// Data to be stored as checkpoint
+type Data struct {
+	Pod      *v1.Pod
+	CheckSum uint64
```
Comment (cmluciano) [inline pkg/kubelet/checkpoint/checkpoint.go]: Checksum is camel-case for this struct but the functions are cased as UpdateChecksum. Can we change this field to just Checksum?

**Comment by vikaschoudhary16** [comment]
ping @vishh @derekwaynecarr @yujuhong @sjenning @dashpole @dchen1107

File: pkg/kubelet/volumemanager/volume_manager_test.go [hunk 2]
```diff
@@ -30,10 +30,11 @@ import (
 	"k8s.io/client-go/kubernetes/fake"
 	"k8s.io/client-go/tools/record"
 	utiltesting "k8s.io/client-go/util/testing"
+	//"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
 	"k8s.io/kubernetes/pkg/kubelet/config"
 	"k8s.io/kubernetes/pkg/kubelet/configmap"
 	containertest "k8s.io/kubernetes/pkg/kubelet/container/testing"
-	"k8s.io/kubernetes/pkg/kubelet/pod"
+	//	"k8s.io/kubernetes/pkg/kubelet/pod"
```
Comment (sjenning) [inline pkg/kubelet/volumemanager/volume_manager_test.go]: nit: remove this.  could be done in a follow-on though.

**Comment by dchen1107** [comment]
Discussed this at today's sig-node, and thanks for refactor the code. Could you please add a README file which points to a doc or tracks all the checkpoint done on the node?

**Comment by jiayingz** [comment]
@vikaschoudhary16 could you rebase this PR and add a README? Thanks!

File: pkg/kubelet/cm/deviceplugin/manager.go [hunk 3]
```diff
@@ -432,56 +431,39 @@ func (m *ManagerImpl) GetCapacity() (v1.ResourceList, v1.ResourceList, []string)
...
 
 // Reads device to container allocation information from disk, and populates
 // m.allocatedDevices accordingly.
 func (m *ManagerImpl) readCheckpoint() error {
-	content, err := m.store.Read(kubeletDevicePluginCheckpoint)
+	checkpoint := NewDevicePluginCheckpoint()
+	err := m.checkpointManager.GetCheckpoint(kubeletDevicePluginCheckpoint, checkpoint)
 	if err != nil {
-		if err == utilstore.ErrKeyNotFound {
+		if err == errors.ErrCheckpointNotFound {
```
Comment (kdembler) [inline pkg/kubelet/cm/deviceplugin/manager.go]: As far as I can see this error is defined and is checked for but isn't actually returned by any function.
In case of nonexistent checkpoint, `GetCheckpoint()` will return `ErrKeyNotFound` from `utilstore` package.
So I guess either `GetCheckpoint()` should check for `ErrKeyNotFound` and return `ErrCheckpointNotFound` or `ErrCheckpointNotFound` should be dropped.
In case of `deviceplugin` it doesn't make a big difference as error returned by `NewDevicePluginCheckpoint` is ignored and only logged anyway, but it may cause inconsistency and errors in other packages.

Comment (iwankgb) [reply pkg/kubelet/cm/deviceplugin/manager.go]: @sjenning do you think it should be addressed before merging or rather in a follow-up PR?
CC: @ConnorDoyle

Comment (vikaschoudhary16) [reply pkg/kubelet/cm/deviceplugin/manager.go]: Updated :)
And rebased as well. Thanks!!

**Comment by vikaschoudhary16** [comment]
ping @dchen1107 @derekwaynecarr @vishh @sjenning @jiayingz 
Updated README as was discussed in the meeting. Sorry for a long delay:
https://github.com/kubernetes/kubernetes/pull/56040/commits/f94268fe241829d6add747f7625de5c3f569abfc#diff-875c8d813e30fd51bced9baf89f8eee1

File: pkg/kubelet/checkpoint/checkpoint.go [hunk 4]
```diff
@@ -39,54 +36,57 @@ const (
...
+	return uint64(hash.Sum32())
 }
 
-// GetInstance will return the current Manager, there should be only one.
-func GetInstance() Manager {
-	mutex.Lock()
-	defer mutex.Unlock()
-	return instance
+// VerifyChecksum verifies that passed checksum is same as calculated checksum
+func (cp *Data) VerifyChecksum() error {
```
Comment (dashpole) [inline pkg/kubelet/checkpoint/checkpoint.go]: It doesn't look like this is used.

File: pkg/kubelet/checkpointmanager/checkpoint_manager.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	utilstore "k8s.io/kubernetes/pkg/kubelet/util/store"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// Checkpoint provides the process checkpoint data
+type Checkpoint interface {
+	MarshalCheckpoint() ([]byte, error)
+	UnmarshalCheckpoint(blob []byte) error
+	GetChecksum() uint64
+	UpdateChecksum()
```
Comment (dashpole) [inline pkg/kubelet/checkpointmanager/checkpoint_manager.go]: I think we can include updating the checksum as part of the MarshalCheckpoint function.

File: pkg/kubelet/checkpoint/checkpoint.go [hunk 6]
```diff
@@ -39,54 +36,57 @@ const (
...
-// NewCheckpointManager will create a Manager that points to the following path
-func NewCheckpointManager(path string) Manager {
-	// NOTE: This is a precaution; current implementation should not run
-	// multiple checkpoint managers.
-	mutex.Lock()
-	defer mutex.Unlock()
-	instance = &fileCheckPointManager{path: path}
-	return instance
+// GetChecksum returns calculated checksum of checkpoint data
+func (cp *Data) GetChecksum() uint64 {
```
Comment (dashpole) [inline pkg/kubelet/checkpoint/checkpoint.go]: I dont think this needs to be an exported method.

File: pkg/kubelet/checkpoint/checkpoint_test.go [hunk 7]
```diff
@@ -25,6 +25,8 @@ import (
 	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
+	//podtest "k8s.io/kubernetes/pkg/kubelet/pods/testing"
```
Comment (dashpole) [inline pkg/kubelet/checkpoint/checkpoint_test.go]: remove unused import.

File: pkg/kubelet/dockershim/docker_checkpoint.go [hunk 8]
```diff
@@ -65,89 +62,47 @@ type PodSandboxCheckpoint struct {
...
+	return uint64(hash.Sum32())
 }
 
-func NewPodSandboxCheckpoint(namespace, name string) *PodSandboxCheckpoint {
-	return &PodSandboxCheckpoint{
-		Version:   schemaVersion,
-		Namespace: namespace,
-		Name:      name,
-		Data:      &CheckpointData{},
+func (cp *PodSandboxCheckpoint) VerifyChecksum() error {
```
Comment (dashpole) [inline pkg/kubelet/dockershim/docker_checkpoint.go]: Remove this since it isn't used.

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager/errors"
+	utilstore "k8s.io/kubernetes/pkg/kubelet/checkpointmanager/testing"
+	hashutil "k8s.io/kubernetes/pkg/util/hash"
+)
+
+type Protocol string
```
Comment (dashpole) [inline pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: Can you add a comment that this struct is used as an example of a struct that can be checkpointed.  We should also not export these structs if that is what is used for.

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+func (cp *CheckpointData) GetChecksum() uint64 {
+	orig := cp.Checksum
+	cp.Checksum = 0
+	hash := fnv.New32a()
+	hashutil.DeepHashObject(hash, *cp)
+	cp.Checksum = orig
+	return uint64(hash.Sum32())
+}
+
+func (cp *CheckpointData) VerifyChecksum() error {
```
Comment (dashpole) [inline pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: This doesnt appear to be used.

File: pkg/kubelet/cm/devicemanager/checkpoint.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	orig := cp.Checksum
+	cp.Checksum = 0
+	hash := fnv.New32a()
+	hashutil.DeepHashObject(hash, *cp)
+	cp.Checksum = orig
+	return uint64(hash.Sum32())
+}
+
+// VerifyChecksum verifies that passed checksum is same as calculated checksum
+func (cp *checkpointData) VerifyChecksum() error {
```
Comment (dashpole) [inline pkg/kubelet/cm/devicemanager/checkpoint.go]: not used.

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 12]
```diff
@@ -432,56 +431,39 @@ func (m *ManagerImpl) GetCapacity() (v1.ResourceList, v1.ResourceList, []string)
...
-
 // Checkpoints device to container allocation information to disk.
 func (m *ManagerImpl) writeCheckpoint() error {
 	m.mutex.Lock()
-	data := checkpointData{
-		PodDeviceEntries:  m.podDevices.toCheckpointData(),
-		RegisteredDevices: make(map[string][]string),
-	}
+	data := NewDevicePluginCheckpoint()
+	data.(*checkpointData).podDeviceEntries = m.podDevices.toCheckpointData()
```
Comment (dashpole) [inline pkg/kubelet/cm/devicemanager/manager.go]: can we pass podDevices and healthyDevices (or whatever we need) to NewDevicePluginCheckpoint rather than casting to checkpointData to reveal non-exported fields?

File: pkg/kubelet/cm/devicemanager/pod_devices.go [hunk 13]
```diff
@@ -21,6 +21,7 @@ import (
 
 	"k8s.io/apimachinery/pkg/util/sets"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+	//	"k8s.io/kubernetes/pkg/kubelet/checkpointmanager"
```
Comment (dashpole) [inline pkg/kubelet/cm/devicemanager/pod_devices.go]: remove commented out import

**Review by dashpole — COMMENTED** [review]
first pass

File: pkg/kubelet/checkpoint/checkpoint.go (hunk 4 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpoint/checkpoint.go]: Thanks for catching this. Intention is to use it from checkpoint manager's ReadCheckpoint()

File: pkg/kubelet/checkpointmanager/checkpoint_manager.go (hunk 5 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: Reason for keeping it separate is to make checksum as part of common checkpoint-manager framework and keeping it explicitly in the interface to ensure that all implementers are going to implement checksum functionality for the sake of consistency. Same is the case with GetChecksum and VerifyChecksum. These checksum related apis can be removed from interface and then an implementer submodule may or may not use checksums.
 
What you  guys think?

File: pkg/kubelet/checkpoint/checkpoint.go (hunk 6 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpoint/checkpoint.go]: reason for this explained above.

File: pkg/kubelet/checkpoint/checkpoint_test.go (hunk 7 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpoint/checkpoint_test.go]: Thanks!

File: pkg/kubelet/dockershim/docker_checkpoint.go (hunk 8 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/dockershim/docker_checkpoint.go]: will use it in checkpoint manager.

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go (hunk 10 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: explained above.

File: pkg/kubelet/cm/devicemanager/checkpoint.go (hunk 11 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/cm/devicemanager/checkpoint.go]: explained above

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 12 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/cm/devicemanager/manager.go]: sure.

File: pkg/kubelet/cm/devicemanager/pod_devices.go (hunk 13 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/cm/devicemanager/pod_devices.go]: sure

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go (hunk 9 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: done.

**Comment by vikaschoudhary16** [comment]
@dashpole Thanks for the comments. I have replied to some of them inline and addressed others. Looking forward for further comments.

ping @sjenning @derekwaynecarr @vishh @dchen1107

File: pkg/kubelet/checkpointmanager/checkpoint_manager.go (hunk 5 above)
Comment (dashpole) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: WDYT about implementing a seperate `struct ChecksumCheckpoint{data interface{}, checksum uint64}` that implements Checkpoint, and handles the checksumming?  Then each library that wants to implement it can just use a wrapper around ChecksumCheckpoint to convert from interface{} to whatever type they need.

Comment (dashpole) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: I think more generally, checksumming shouldn't be something that is leaked to implementations.  It is an implementation detail of Reading and Writing a checkpoint from disk.

Comment (vikaschoudhary16) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: true. OTOH, Since each implementation has different checkpoint data, it is not possible for manager to handle checksum without depending(leaking) on implementers.

Comment (dashpole) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: Makes sense.  I didn't realize json marshalling/unmarshalling doesn't work well with interface{}... which means you can't have a gereral checkpointer that does marshalling/unmarshalling and checkpointing for all implementations.

**Comment by dchen1107** [comment]
Sorry for missing the 1.10 code freeze date. But if we handle well, we can get this pr in once code freeze is lift.

File: pkg/kubelet/checkpointmanager/checkpoint_manager.go (hunk 5 above)
Comment (dashpole) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: I think it is enough to include UpdateChecksum() and VerifyChecksum() in the API, as both of those are used.

**Comment by dchen1107** [comment]
The pr looks good to me. I am going to approve it, but hope @sjenning @yujuhong @jiayingz who introduced checkpoints to kubelet for various components can take a look. 

/approve

File: pkg/kubelet/checkpointmanager/checkpoint_manager.go (hunk 5 above)
Comment (dashpole) [reply pkg/kubelet/checkpointmanager/checkpoint_manager.go]: Take a look at https://github.com/kubernetes/kubernetes/commit/5bde42a9177627999d6cee00c0768228425dde49.  I broke out the checksumming into its own package, and avoid casting to types by adding PodCheckpoint and DeviceManagerCheckpoint interfaces.

**Review by jiayingz — COMMENTED** [review]
Thanks a lot for making the change! I made a few nit comments. Overall looks great!

File: pkg/kubelet/dockershim/docker_checkpoint.go [hunk 14]
```diff
@@ -65,89 +65,31 @@ type PodSandboxCheckpoint struct {
...
-	// WARNING: RemoveCheckpoint will not return error if checkpoint does not exist.
-	RemoveCheckpoint(podSandboxID string) error
-	// ListCheckpoint returns the list of existing checkpoints.
-	ListCheckpoints() ([]string, error)
-}
-
-// PersistentCheckpointHandler is an implementation of CheckpointHandler. It persists checkpoint in CheckpointStore
-type PersistentCheckpointHandler struct {
-	store utilstore.Store
+	Checksum checksum.Checksum `json:"checksum"`
```
Comment (dashpole) [inline pkg/kubelet/dockershim/docker_checkpoint.go]: is Version used anywhere?  Should we be checking this?

File: pkg/kubelet/dockershim/docker_checkpoint.go [hunk 15]
```diff
@@ -65,89 +65,31 @@ type PodSandboxCheckpoint struct {
...
-	if err != nil {
-		glog.Errorf("Failed to unmarshal checkpoint %q, removing checkpoint. Checkpoint content: %q. ErrMsg: %v", podSandboxID, string(blob), err)
-		handler.RemoveCheckpoint(podSandboxID)
-		return nil, fmt.Errorf("failed to unmarshal checkpoint")
-	}
-	if checkpoint.CheckSum != calculateChecksum(checkpoint) {
-		glog.Errorf("Checksum of checkpoint %q is not valid, removing checkpoint", podSandboxID)
-		handler.RemoveCheckpoint(podSandboxID)
-		return nil, fmt.Errorf("checkpoint is corrupted")
+func NewPodSandboxCheckpoint(namespace, name string) DockershimCheckpoint {
```
Comment (dashpole) [inline pkg/kubelet/dockershim/docker_checkpoint.go]: Can we set PortMappings and HostNetwork here as well?

File: pkg/kubelet/dockershim/docker_checkpoint.go [hunk 16]
```diff
@@ -65,89 +65,31 @@ type PodSandboxCheckpoint struct {
...
+	return cp.Checksum.Verify(*cp.Data)
 }
 
-func calculateChecksum(checkpoint PodSandboxCheckpoint) uint64 {
-	checkpoint.CheckSum = 0
-	hash := fnv.New32a()
-	hashutil.DeepHashObject(hash, checkpoint)
-	return uint64(hash.Sum32())
+func (cp *PodSandboxCheckpoint) GetData() ([]*PortMapping, bool) {
+	return cp.Data.PortMappings, cp.Data.HostNetwork
```
Comment (dashpole) [inline pkg/kubelet/dockershim/docker_checkpoint.go]: Can we return Name and Namespace here as well?

Comment (dashpole) [reply pkg/kubelet/dockershim/docker_checkpoint.go]: Or add GetMeta() or something to return those if that makes more sense

File: pkg/kubelet/cm/devicemanager/manager_test.go [hunk 17]
```diff
@@ -302,21 +301,56 @@ func constructAllocResp(devices, mounts, envs map[string]string) *pluginapi.Cont
...
+
 func TestCheckpoint(t *testing.T) {
 	resourceName1 := "domain1.com/resource1"
 	resourceName2 := "domain2.com/resource2"
-
 	as := assert.New(t)
-	tmpDir, err := ioutil.TempDir("", "checkpoint")
-	as.Nil(err)
-	defer os.RemoveAll(tmpDir)
+	ckm := newMockCheckpointManager()
```
[the remark this answers is not in the document]
Comment (vikaschoudhary16) [reply pkg/kubelet/cm/devicemanager/manager_test.go]: Unit tests are expected to test specific code logic in isolation from runtime-dependencies and cross-package dependencies. mockCheckpointManager is implementing the checkpointManager interface, therefore unit tests are testing device manager logic in isolation but verifying the API contract with checkpointManager.

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+		assert.NoError(t, err)
+		actualPortMappings, actualHostNetwork := checkpointOut.GetData()
+		expPortMappings, expHostNetwork := tc.checkpoint.GetData()
+		assert.Equal(t, actualPortMappings, expPortMappings)
+		assert.Equal(t, actualHostNetwork, expHostNetwork)
+	}
+	// Test it fails for reading V1 into V2
+	checkpointV2 := newFakeCheckpointV2("", nil, false)
+	err = manager.GetCheckpoint("key1", checkpointV2)
+	assert.EqualError(t, err, "checkpoint is corrupted")
```
Comment (jiayingz) [inline pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: Just to make sure I understand, the reason GetCheckpoint fails here is because we use a different data struct for the field? If we just add a new field in the checkpoint data, reading old checkpoint data without that field should still succeed, right?

File: pkg/kubelet/cm/devicemanager/manager_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (jiayingz) [reply pkg/kubelet/cm/devicemanager/manager_test.go]: Should we provide a fake impl in checkpointmanager pkg to avoid duplicate code in individual components?

**Review by jiayingz — COMMENTED** [review]
Thanks a lot for the changes!

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go (hunk 18 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: right. But if you see updated changes, even if i read the checkpoint into a same structure i.e same name and same members (but defined in different package), it fails at checksum verification. I was expecting it to pass. Also if i delete a member while reading, still it fails at checksum rather failing at unmarshalling. Rest of your comments have been addressed, PTAL.

Comment (jiayingz) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: I see. I think (also verified manually) if we keep the same field name and struct name, but just adding a new field, GetCheckpoint of the old data without the field will succeed. Unfortunately this seems hard to test in a unit test. Also, at least for devicemanager, upgrading to this new version will cause reading previous checkpoint files to fail. There is a potential to hit the similar race that https://github.com/kubernetes/kubernetes/pull/60856 fixes if someone is trying to do in-place kubelet upgrade. Given that in most k8s environment, upgrading kubelet involves deleting or draining the old-version node, and for devicemanager, the failure should be recoverable, it is probably not worth adding extra code complexity to make sure we can read old version of checkpoint file. But I think you should add a release note to clear mention this.

@Random-Liu and @yujuhong could you also check whether this change could cause checkpoint read failure for other kubelet components?

File: pkg/kubelet/checkpointmanager/checkpoint_manager_test.go [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+		assert.NoError(t, err)
+		actualPortMappings, actualHostNetwork := checkpointOut.GetData()
+		expPortMappings, expHostNetwork := tc.checkpoint.GetData()
+		assert.Equal(t, actualPortMappings, expPortMappings)
+		assert.Equal(t, actualHostNetwork, expHostNetwork)
+	}
+	// Test it fails if tried to read V1 structure into V2, a different structure from the structure which is checkpointed
+	checkpointV2 := newFakeCheckpointV2("", nil, false)
+	err = manager.GetCheckpoint("key1", checkpointV2)
+	assert.EqualError(t, err, "checkpoint is corrupted")
```
Comment (jiayingz) [inline pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: I spent some time looking into this today. I think the reason checksum checking fails here is because the hashutil.DeepHashObject() function used in getChecksum() uses spew printer that also prints out the field type. Here are what it generates from spew printer.Fprintf:

for checkpoint1:
(checkpointmanager.Data){PortMappings:([]*checkpointmanager.PortMapping)[<*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)80 HostPort:(*int32)80} <*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)443 HostPort:(*int32)443}] HostNetwork:(bool)true}

for checkpoint2:
(checkpointmanager.CheckpointDataV2){PortMappings:([]*checkpointmanager.PortMapping)[<*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)80 HostPort:(*int32)80} <*>{Protocol:(*checkpointmanager.protocol)tcp ContainerPort:(*int32)443 HostPort:(*int32)443}] HostNetwork:(bool)true V2Field:(string)}

Basically, this means that checksum checking would always fail whenever we add new fields to the checkpointed data structure, even though json Unmarshal would pass. Looking at the change, I did notice that this is also the current behavior in docker_checkpoint.go, but @yujuhong @Random-Liu and @dchen1107 can you verify that this is indeed the desired behavior for kubelet?

Comment (dchen1107) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: Good point! Today we shouldn't run into any issue since we are doing a bundled upgrade on a node, hence there wouldn't have a new field be introduced to the checkpointed data structure when kubelet is restarted without tearing-down-then-recreation. But once we introduced in-place-upgrade, we might encounter the issue. @yujuhong is it possible to change such behavior in docker-shim's checkpoints? Or we want to keep that to prevent arbitrary checkpointed data structure changes?

Comment (yujuhong) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: The purpose of the checksum is to detect corruption. I *think* with dockershim's checkpoint, you're supposed to version the checkpoint struct, and adding/changing a field requires a version bump. Alternatively, you can also hash the raw content of the file, but that requires saving the checksum as a separate file. +@freehan (the author of dockershim's checkpointing) in case I'm wrong.

Comment (vikaschoudhary16) [reply pkg/kubelet/checkpointmanager/checkpoint_manager_test.go]: @jiayingz Done, PTAL!!

**Comment by jiayingz** [comment]
For the devicemanager part of change:
/lgtm
/approve
